File: php/sdk/otel/instrumentation.php
<?php

namespace OtelInstrumentation;

use OpenTelemetry\Contrib\Grpc\GrpcTransportFactory;
use OpenTelemetry\Contrib\Otlp\OtlpUtil;
use OpenTelemetry\Contrib\Otlp\LogsExporter;
use OpenTelemetry\SDK\Logs\LoggerProvider;
use OpenTelemetry\API\Signals;
use OpenTelemetry\Contrib\Otlp\SpanExporter;
use OpenTelemetry\SDK\Trace\SpanProcessor\SimpleSpanProcessor;
use OpenTelemetry\SDK\Trace\TracerProvider;
use OpenTelemetry\SDK\Logs\Processor\SimpleLogRecordProcessor;
use OpenTelemetry\SDK\Common\Attribute\Attributes;
use OpenTelemetry\SDK\Logs\Exporter\ConsoleExporterFactory;
use Illuminate\Support\Facades\Log;
use OpenTelemetry\Contrib\Otlp\MetricExporter;
use OpenTelemetry\SDK\Metrics\Data\Temporality;
use OpenTelemetry\SDK\Metrics\MeterProvider;
use OpenTelemetry\SDK\Metrics\MetricReader\ExportingReader;
use OpenTelemetry\SDK\Resource\ResourceInfoFactory;
use OpenTelemetry\SDK\Common\Export\Stream\StreamTransportFactory;
use OpenTelemetry\SemConv\ResourceAttributes;
use OpenTelemetry\SDK\Resource\ResourceInfo;
use OpenTelemetry\SDK\Sdk;
use OpenTelemetry\API\Trace\Propagation\TraceContextPropagator;
use OpenTelemetry\SDK\Trace\Sampler\AlwaysOnSampler;
use OpenTelemetry\SDK\Trace\Sampler\ParentBased;


class Instrumentation
{
    public $tracer;
    public $logger;
    public $metrics;
    public $resource;
    public $otelEndpoint;

    public function __construct()
    {
        $this->otelEndpoint = isset($_ENV['OTEL_EXPORTER_OTLP_ENDPOINT']) ? $_ENV['OTEL_EXPORTER_OTLP_ENDPOINT'] : 'http://localhost:4317';
        $this->resource = ResourceInfoFactory::emptyResource()->merge(ResourceInfo::create(Attributes::create([
            ResourceAttributes::SERVICE_NAMESPACE => 'otel',
            ResourceAttributes::SERVICE_NAME => getenv('OTEL_SERVICE_NAME') ?? 'otel-sdk-examples-php-sdk',
            ResourceAttributes::SERVICE_VERSION => '0.1.0',
        ])));
        $this->setTracer();
        $this->setLogger();
        $this->setMetrics();
        $this->registerSdk();
    }

    public function setMetrics()
    {
        // Set up metrics
        if (getenv('APP_ENV') == 'dev') {
            $reader = new ExportingReader(new MetricExporter((new StreamTransportFactory())->create(fopen('php://stdout', 'wb'), 'application/x-ndjson'), /*Temporality::CUMULATIVE*/));

            $meterProvider = MeterProvider::builder()
                ->addReader($reader)
                ->build();

            return $this->metrics = $meterProvider;
        } else {
            $transport = (new GrpcTransportFactory())->create($this->otelEndpoint . OtlpUtil::method(Signals::METRICS));
            $reader = new ExportingReader(new MetricExporter($transport, Temporality::CUMULATIVE));

            $meterProvider = MeterProvider::builder()
                ->setResource($this->resource)
                ->addReader($reader)
                ->build();

            return $this->metrics = $meterProvider;
        }
    }

    public function setTracer()
    {
        if (getenv('APP_ENV') == 'dev') {
            $tracer = new SpanExporter((new StreamTransportFactory())->create(fopen('php://stdout', 'wb'), 'application/x-ndjson'));

            $tracerProvider = TracerProvider::builder()
                ->addSpanProcessor(
                    new SimpleSpanProcessor($tracer)
                )
                ->setResource($this->resource)
                ->setSampler(new ParentBased(new AlwaysOnSampler()))
                ->build();

            return $this->tracer = $tracerProvider;
        } else {
            $transport = (new GrpcTransportFactory())->create($this->otelEndpoint . OtlpUtil::method(Signals::TRACE));
            $spanExporter = new SpanExporter($transport);

            $tracerProvider = TracerProvider::builder()
                ->addSpanProcessor(
                    new SimpleSpanProcessor($spanExporter)
                )
                ->setResource($this->resource)
                ->setSampler(new ParentBased(new AlwaysOnSampler()))
                ->build();

            return $this->tracer = $tracerProvider;
        }
    }

    public function setLogger()
    {
        // Set up logging
        if (getenv('APP_ENV') == 'dev') {
            Log::channel('stdout')->info(message: 'OTEL endpoint: ' . $this->otelEndpoint);

            $loggerProvider = LoggerProvider::builder()
                ->setResource($this->resource)
                ->addLogRecordProcessor(
                    new SimpleLogRecordProcessor(
                        (new ConsoleExporterFactory())->create()
                    )
                )
                ->build();

            return $this->logger = $loggerProvider;
        } else {
            Log::channel('stdout')->info(message: 'OTEL endpoint: ' . $this->otelEndpoint);

            $transport = (new GrpcTransportFactory())->create($this->otelEndpoint . OtlpUtil::method(Signals::LOGS));
            $logExporter = new LogsExporter(
                $transport
            );

            // Create LoggerProvider for logs
            $loggerProvider = LoggerProvider::builder()
                ->setResource($this->resource)
                ->addLogRecordProcessor(
                    new SimpleLogRecordProcessor($logExporter)
                )
                ->build();

            return $this->logger = $loggerProvider;
        }
    }

    public function registerSdk()
    {
        // Build and register the global SDK
        Sdk::builder()
            ->setTracerProvider($this->tracer)
            ->setMeterProvider($this->metrics)
            ->setLoggerProvider($this->logger)
            ->setPropagator(TraceContextPropagator::getInstance())
            ->setAutoShutdown(true)
            ->buildAndRegisterGlobal();
    }

    public function getTracer()
    {
        return $this->tracer;
    }

    public function getLogger()
    {
        return $this->logger;
    }

    public function getMetrics()
    {
        return $this->metrics;
    }
}

File: python/sdk/instrumentation.py
import logging
import os

from opentelemetry import metrics, trace
from opentelemetry._logs import set_logger_provider
from opentelemetry.exporter.otlp.proto.grpc._log_exporter import \
    OTLPLogExporter
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import \
    OTLPMetricExporter
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import \
    OTLPSpanExporter
from opentelemetry.sdk._logs import LoggerProvider, LoggingHandler
from opentelemetry.sdk._logs.export import (BatchLogRecordProcessor,
                                            ConsoleLogExporter)
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import (ConsoleMetricExporter,
                                              PeriodicExportingMetricReader)
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import (BatchSpanProcessor,
                                            ConsoleSpanExporter)


def initialize_instrumentation():
    # Service name is required for most backends
    resource = Resource(attributes={
        SERVICE_NAME: os.getenv('OTEL_SERVICE_NAME',
                                'otel-sdk-examples-python-sdk')
    })

    traceProvider = TracerProvider(resource=resource)

    if os.getenv('ENVIRONMENT') == 'dev':
        processor = BatchSpanProcessor(ConsoleSpanExporter())
    else:
        processor = BatchSpanProcessor(OTLPSpanExporter(
            endpoint=os.getenv('OTEL_EXPORTER_OTLP_ENDPOINT', 'localhost:4317')))

    traceProvider.add_span_processor(processor)
    trace.set_tracer_provider(traceProvider)

    if os.getenv('ENVIRONMENT') == 'dev':
        reader = PeriodicExportingMetricReader(ConsoleMetricExporter())
    else:
        reader = PeriodicExportingMetricReader(
            OTLPMetricExporter(endpoint=os.getenv(
                'OTEL_EXPORTER_OTLP_ENDPOINT', 'localhost:4317'))
        )

    meterProvider = MeterProvider(resource=resource, metric_readers=[reader])
    metrics.set_meter_provider(meterProvider)

def initialize_otel_logging():
    logger_provider = LoggerProvider(
        resource=Resource.create(
            {
                "service.name": os.getenv('OTEL_SERVICE_NAME',
                                'otel-sdk-examples-python-sdk'),
            }
        ),
    )
    set_logger_provider(logger_provider)

    if os.getenv('ENVIRONMENT') == 'dev':
        exporter = ConsoleLogExporter()
    else:
        exporter = OTLPLogExporter(insecure=True)

    logger_provider.add_log_record_processor(BatchLogRecordProcessor(exporter))
    handler = LoggingHandler(level=logging.DEBUG, logger_provider=logger_provider)

    # Attach OTLP handler to root logger
    logging.getLogger().addHandler(handler)
    logging.getLogger().addHandler(logging.StreamHandler())
    logging.basicConfig(level=logging.INFO)

    logging.getLogger('instrumentation').setLevel(logging.DEBUG)
    app_logger = logging.getLogger('instrumentation')
    app_logger.info("Logging initialized")

    return app_logger

